torch/lib/c10d/TCPStore.hpp

```diff
@@ -36,29 +70,58 @@ class TCPStoreDaemon {
   void getNumKeysHandler(int socket) const;
   void deleteHandler(int socket);
   void waitHandler(int socket);
+  void watchHandler(int socket);
 
   bool checkKeys(const std::vector<std::string>& keys) const;
+  // Helper function to alerts waiting workers, used in setHandler, getHandler
   void wakeupWaitingClients(const std::string& key);
-
-  void initStopSignal();
-  void closeStopSignal();
-
-  std::thread daemonThread_;
+  // Helper function used when the key is changed
+  // used in setHandler, addHandler, getHandler, deleteHandler
+  void sendKeyUpdatesToClients(
+      const std::string& key,
+      const enum WatchResponseType& type,
+      std::vector<uint8_t>& oldData,
+      std::vector<uint8_t>& newData);
   std::unordered_map<std::string, std::vector<uint8_t>> tcpStore_;
-  // From key -> the list of sockets waiting on it
+  // From key -> the list of sockets waiting on the key
   std::unordered_map<std::string, std::vector<int>> waitingSockets_;
   // From socket -> number of keys awaited
   std::unordered_map<int, size_t> keysAwaited_;
+  // From key -> the list of sockets watching the key
+  std::unordered_map<std::string, std::vector<int>> watchedSockets_;
+};
 
-  std::vector<int> sockets_;
-  int storeListenSocket_;
-#ifdef _WIN32
-  const std::chrono::milliseconds checkTimeout_
-      = std::chrono::milliseconds(10);
-  HANDLE ghStopEvent_;
-#else
-  std::vector<int> controlPipeFd_{-1, -1};
-#endif
+// Separate thread that is launched on all instances (including master)
+// Right now only handles callbacks registered from watchKey()
+class TCPStoreWorkerDaemon : public BackgroundThread {
+ public:
+  explicit TCPStoreWorkerDaemon(int listenSocket);
+  // Set the callback to run key change
+  void setCallback(std::string key, WatchKeyCallback cb);
+  void waitForCallbackRegistration() {
+    // Block until callback has been registered successfully
+    std::unique_lock<std::mutex> callbackRegistrationLock(
+        callbackRegistrationMutex_);
+    callbackRegisteredCV_.wait(
+        callbackRegistrationLock, [&] { return callbackRegisteredData_; });
+
+    // Reset payload for next callback
+    callbackRegisteredData_ = false;
+  }
+  void setCallbackRegistered() {
+    callbackRegisteredData_ = true;
```

Comment:
do we need `callbackRegistrationMutex_` to protect this set here? I guess it is not required today, because `TCPStore::watchKey` will block until `callbackRegisteredData_` is set to `True`. So the second watchKey request won't be sent until the `nofity_one()` below is called?

Reply:
Yep, that is correct. Not required since only one watchKey request can be sent at a time.